src/lib/frontend/models.ml

```diff
@@ -432,41 +455,48 @@ module Why3CounterExample = struct
 
   let output_constraints fmt prop_model =
     let assertions = SE.fold (fun e acc ->
-        (asprintf "%s(assert %a)@ " acc SmtlibCounterExample.pp_term e)
-      ) prop_model "" in
-    Printer.print_fmt ~flushed:false fmt "@ ; constants@ ";
+        (dprintf "%t(assert %a)@ " acc SmtlibCounterExample.pp_term e)
+      ) prop_model (dprintf "") in
+    Format.fprintf fmt "@ ; constraints@ ";
     MS.iter (fun _ (name,ty,args_ty) ->
         match args_ty with
         | [] ->
-          Printer.print_fmt ~flushed:false fmt "(declare-const %s %s)@ "
+          Format.fprintf fmt "(declare-const %s %s)@ "
             name ty
         | l ->
-          Printer.print_fmt ~flushed:false fmt "(declare-fun %s (%s) %s)@ "
+          Format.fprintf fmt "(declare-fun %s (%s) %s)@ "
             name
             (String.concat " " l)
             ty
       ) !constraints;
-    Printer.print_fmt ~flushed:false fmt "@ ; assertions@ ";
-    Printer.print_fmt fmt ~flushed:false "%s" assertions
+    Format.fprintf fmt "@ ; assertions@ ";
+    Format.fprintf fmt "%t" assertions
 
 end
 (* of module Why3CounterExample *)
 
 let output_concrete_model fmt props ~functions ~constants ~arrays =
-  Printer.print_fmt ~flushed:false fmt "@[<v 2>(@,";
+  Format.fprintf fmt "@[<v 2>(";
   if Options.get_model_type_constraints () then
     begin
-      Why3CounterExample.output_constraints fmt props
+      Why3CounterExample.output_constraints fmt props;
+      Format.fprintf fmt "@ ; values"
     end;
 
-  Printer.print_fmt fmt "@ ; Functions@ ";
+  (* Functions *)
   let records = SmtlibCounterExample.output_functions_counterexample
       fmt  MS.empty functions in
 
-  Printer.print_fmt fmt "@ ; Constants@ ";
+  let array_symbols =
+    ModelMap.fold (fun (f, _, _) _ -> Sy.Set.add f) arrays Sy.Set.empty
+  in
+
+  (* Constants *)
+  (* If the constant is present in the arrays map, it will be printed there. *)
```

Comment:
If I remember correctly, this is because arrays for which there has been no get/set are not added to the `array_symbols`.

I didn't think it was very important to keep this distinction, but it should not be too hard to iterate over the constant array symbols in `output_arrays_counterexample` instead.

Reply:
I have a version that does what I described above, but it's not very pretty because:

 - There are arrays that only appear in the `constants` portion. This is the case where an array is defined to be equal to another array.
 - There are *also* arrays that only appear in the `arrays` portion. This is the case of an array that is defined (i.e. has get/set on them) but does *not* appear in the original problem and was synthesized by the solver.

Here is a (slightly overcomplicated) example:

```smtlib2
(set-logic ALL)
(declare-const x Int)
(declare-const y Int)
(declare-const a1 (Array Int Int))
(declare-const a2 (Array Int Int))
(declare-const a4 (Array Int Int))
(assert (= a4 a2))
(assert (= (select a1 x) x))
(assert (= (store a1 x y) a1))
(assert (= (store a1 y y) a1))
(assert (= (store a2 7 3) a2))
(assert (= (store a2 19 4) a2))

(declare-const a3 (Array Int Int))

(declare-fun f (Bool) (Array Int Int))
(assert (= (f true) a2))
(assert (= (store (f false) 7 3) (f false)))

(check-sat)
(get-model)
```

A model is:

```
unknown
(
  ; Functions
  (define-fun f ((arg_0 Bool)) (Array Int Int) (ite (= arg_0 true) a4 !k2))
  ; Constants
  (define-fun x () Int 0)
  (define-fun y () Int 0)
  ; Arrays
  (define-fun a1 () (Array Int Int) (store (as @a1 (Array Int Int)) 0 0))
  (define-fun a2 () (Array Int Int) a4)
  (define-fun a4 () (Array Int Int)
   (store (store (as @a4 (Array Int Int)) 7 3) 19 4))
  (define-fun !k2 () (Array Int Int) (store (as @!k2 (Array Int Int)) 7 3))
)
```

In this example, `a1` and `a4` are in both `constants` and `arrays`; `a2` is only in `constants`; and `!k2` is only in `arrays`.

Reply:
(I will push the code but am waiting on the merge of #703 to include tests)

Reply:
That said I am not convinced that keeping the separation while printing is worth the added code complexity.

Reply:
I agree. All the values are functions. We should remove the difference. I suggest to keep the code as simple as possible and remove the pointless SMTLIB comments in the output model. We will produce a better code for the next release.